### PR TITLE
Caching skip-able entries in DDLWorker

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Caching skip-able entries while executing DDL from Zookeeper distributed DDL queue

So we can quickly skip entries without doing some heavy jobs in `DDLWorker::initAndCheckTask`

This PR complements #54819, together improve the DDLWorker speed, especially when machine under high load.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 3646c9fa587739f687fd0030629089320d13c756 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54828/3646c9fa587739f687fd0030629089320d13c756/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Comment by canhld94** [comment]
It seem we don't need to reset the cache when re-init the DDL worker. Once an entry written to ZK, it will not change.

**Comment by canhld94** [comment]
@tavplubix hi, do you think this is useful for `ON CLUSTER` query? Thanks!

File: src/Interpreters/DDLWorker.h [hunk 1]
```diff
@@ -79,6 +81,33 @@ class DDLWorker
...
+
+        bool remove(const String & key)
+        {
+            std::unique_lock lock(mtx);
+            return set.erase(key);
+        }
+
+    private:
+        std::unordered_set<String> set;
+        mutable std::shared_mutex mtx;
```
Comment (tavplubix) [inline src/Interpreters/DDLWorker.h:108]: Shared mutex is overkill, an ordinary mutex would be enough

**Review by tavplubix — APPROVED** [review]
Sorry, I was busy and didn't have time to review it. Yes, the changes make sense.

**Comment by tavplubix** [comment]
@canhld94, could you please update the changelog entry? Currently, it's not user-readable (users have no idea of what `DDLWorker::initAndCheckTask` is)

**Pull request merged**